### DISCUSSION TITLE
Dissallow reimplementing interface members

### DISCUSSION THREAD

**Discussion by helviett** [dotnet/csharplang#6145] [Language Ideas]
In our game engine we always try to adapt suitable features from new C# release. I played with a default interface implementation and found that I need some way to restrict class from implementing interface member that already has default implementation. Consider the following example:

```csharp
public interface IPresenter
{
	void Render(Node node);
	bool PartialHitTest(Node node, ref HitTestArgs args);
}

public interface IWidgetPresenter : IPresenter
{
	void IPresenter.Render(Node node)
	{
		var widget = (Widget)node;
		Renderer.Transform1 = widget.LocalToWorldTransform;
		Renderer.Blending = widget.Blending;
		Renderer.Shader = widget.Shader;
		Render(widget);
	}

	public void Render(Widget widget);
}
```

The default implementation of `IPresenter.Render` in `IWidgetPresenter` does routine work and calls more specific method `Render(Widget widget)`. But unfortunately the user of the interface can implement both `IPresenter.Render(Node)` and `IWidgetPresenter.Render(Widget)` which can lead to problems or misunderstanding. I think that user of interface should be informed that `IWidgetPresenter` has already implemented `IPresenter.Render` and this implementation mustn't be changed because we provide another more specific behaviour. Otherwise user should know internals.

Desired behaviour can be achieved with abstract classes:
```csharp
public abstract class WidgetPresenter : IPresenter
{
	public void Render(Node node) { /* ... */ }
	public abstract bool PartialHitTest(Node node, ref HitTestArgs args);
	public abstract void Render(Widget widget);
}
```
And another way:
```csharp
public abstract class Presenter
{
	public abstract void Render(Node node);
	public abstract bool PartialHitTest(Node node, ref HitTestArgs args);
}

public abstract class WidgetPresenter : Presenter
{
	public sealed override void Render(Node node) { /* ... */ }
	public abstract void Render(Widget widget);
}
```

Perhaps interfaces should use `sealed` keyword in order to prohibit method reimplementation.

**Comment by CyrusNajmabadi** [comment]
Can you check for this and error just using an analyzer?

**Reply by helviett** [reply]
Even if you can users of API must enable analyzer in order to enforce checks.

**Comment by En3Tho** [comment]
Can this be achieved with generic TPresenter dependency on base class? So you basically make like a super base class  specified with this generic dependency and with internal constructor and then use public child classes that will 100% inherit this implementation 

TPresenter can be described as where: new(), IPresenter, so it can be a struct with no fields, for example.

But I guess user defined child classes could still override specific interface implementation...

**Comment by HaloFour** [comment]
Even ignoring DIM there's nothing you can do to prevent a subclass from explicitly implementing `IPresenter` and providing its own explicit implementation of `IPresenter.Render`.

**Reply by helviett** [reply]
> there's nothing you can do to prevent a subclass from explicitly implementing `IPresenter` and providing its own explicit implementation of `IPresenter.Render`

I don't understand why you can't prohibit implementing already implemented members. You create a rule (and probabbly a keyword) and prohibit reimplementation of interface members members (because they already satisfy all needs).

**Reply by HaloFour** [reply]
Because the runtime has no means to prevent it, without that there's nothing that the language can do to enforce how someone might use (or extend) your class.

IMO there's not much reason to want to seal `Render(Node)` in your case.  If all it really does is cast `Node` to `Widget` and then call `Render(Widget)` is it really that problematic that someone might go out of their way to implement it themselves?  If that results in unexpected/undesirable behavior, well, that's what documentation is for.

**Reply by helviett** [reply]
> IMO there's not much reason to want to seal Render(Node) in your case

It also does routine stuff like setting some rendering parameters specific to every widget, so user:

- Won't forget to write this boilerplate, because it is already written;
- Doesn't have to write this boilerplate in every presenter which makes code less complicated.

> is it really that problematic that someone might go out of their way to implement it themselves?

Yes, if there is a way to mess things up then someone will mess them [for sure](https://en.wikipedia.org/wiki/Chekhov%27s_gun). That's why we have access modifiers in every modern programming language for example. In that particular case `abstract class` is flexible enough. I can both prohibit sublcass from overriding `public void Render(Node)` method and mark `Render(Widget)` method as `protected` so noone is able to cast and call the method. 'Sealing' interface method have a semantics of already fulfilled requirement. Same ideas were introduced in trait system in [this paper](http://scg.unibe.ch/archive/papers/Scha03aTraits.pdf).

**Reply by CyrusNajmabadi** [reply]
Feels like you don't want an interface here then.  

Note: i'm not sure why you're against the idea of an analyzer here.  You say it's because someone could ignore the analyzer.  But anyone could ignore this constraint you're adding just by using an existing compiler (or non-c# compiler) that doesn't know anything about this.

This is something interfaces support.  If you don't want that, the best options are an analyzer, or just don't use interfaces here as you want fundamental limits that go against the core ideas of what interfaces (and impls/re-impls) are. :)

**Reply by helviett** [reply]
> Feels like you don't want an interface here then.

I want to avoid inheritance. Inheritance introduces weird mechanism allowing you to call base implementation. IMO it is always not clear where to call base implementation. Should it be in the begining, in the end or somewhere in the middle? You can't answer without having a knowledge of the internals. For example you should always somewhere call base implementation in Window Forms methods overrides in order to trigger associated event. However, you are not able to derive this knowledge from the code, so you have to read source code or documentation. Interfaces don't have this issue that's why I'd like to use them instead of inheritance. By the way class behaviour is easily extendable by implementing another interface but you can't inherit another abstract class (you have to introduce new one, that implements all required interfaces). 

> But anyone could ignore this constraint you're adding just by using an existing compiler (or non-c# compiler) that doesn't know anything about this.

If someone uses compiler that doesn't support a feature it won't be able to compile code. If someone uses some "hacked "c# compiler then he takes all the risk.

**Reply by CyrusNajmabadi** [reply]
> If someone uses compiler that doesn't support a feature it won't be able to compile code.

Interfaces are already supported (as are classes).  So any existing compilers that have shipped since the beginning would still be able to rederive from these types and add reimplementations of those interfaces.
